src/mon/MonClient.h

```diff
@@ -106,7 +105,8 @@ class MonClient : public Dispatcher {
 public:
   MonMap monmap;
 private:
-  MonClientState state;
+  map<entity_addr_t, MonClientState> state_map;
+  map<entity_addr_t, unique_ptr<AuthClientHandler>> auth_map;
```

Comment:
Hmm, I'm not sure about this.  I think you need to hold references to the pending connections.  I think the best way forward would be to define a:

```
struct MonConnection {
  ConnectionRef conn;
  string name;
  unique_ptr<AuthClientHandler> auth;
  MonClientState state;
};
```

Once you have that, you can replace state_map and auth_map with

```
map<entity_inst_t, unique_ptr<MonConnection> > pending_connections;
```

Next, you can replcae curr_conn, cur_mon, and auth with

```
unique_ptr<MonConnection> active_connection;
```

If I'm understanding this right, you can replace hunting and reopen with:

```
bool hunting() const { return !pending_connections.empty(); }
bool reopen() const { return hunting() || active_connection.get() != nullptr; }
```

and at all times

```
assert(pending_connections.empty() || (active_connection.get() == nullptr);
```